marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs

```diff
@@ -0,0 +1,564 @@
+         , tx.id :: bytea
+         , txOut.lovelace :: bigint
+         , txOut.address :: bytea
+    FROM   chain.block AS block
+    JOIN   chain.tx    AS tx    ON tx.blockId = block.id
+                               AND tx.slotNo = block.slotNo
+    JOIN   chain.txOut AS txOut ON txOut.txId   = tx.id
+    WHERE  block.slotNo = -1
+  |]
+  where
+    decodeResults :: [(ByteString, ByteString, Int64, ByteString)] -> Either Text (Maybe GenesisBlock)
+    decodeResults [] = Right Nothing
+    decodeResults rows@((hash, _, _, _) : _) =
+      Just . GenesisBlock (HeaderHash $ toShort hash) . Set.fromList <$> traverse decodeGenesisTx rows
+
+    decodeGenesisTx :: (ByteString, ByteString, Int64, ByteString) -> Either Text GenesisTx
+    decodeGenesisTx (_, txId, lovelace, address) = GenesisTx
+      <$> decodeTxId txId
+      <*> pure (fromIntegral lovelace)
+      <*> decodeAddressAny address
+
+-- GetHeaderAtPoint
+
+getHeaderAtPoint :: GetHeaderAtPoint Session
+getHeaderAtPoint = GetHeaderAtPoint \case
+  ChainPointAtGenesis -> pure Origin
+  point@(ChainPoint slotNo hash) ->
+    statement (slotNoToParam slotNo, headerHashToParam hash) $ refineResult decodeResults
+      [maybeStatement|
+        SELECT block.blockNo :: bigint
+          FROM chain.block AS block
+        WHERE block.slotNo = $1 :: bigint
+          AND block.id     = $2 :: bytea
+      |]
+    where
+      decodeResults :: Maybe Int64 -> Either Text (WithOrigin BlockHeader)
+      decodeResults Nothing        = Left $ "No block found at " <> T.pack (show point)
+      decodeResults (Just blockNo) = Right $ At $ BlockHeader slotNo hash $ BlockNo $ fromIntegral blockNo
+
+decodeTxId :: ByteString -> Either Text TxId
+decodeTxId txId = case deserialiseFromRawBytes AsTxId txId of
+  Nothing    -> Left "Invalid TxId"
+  Just txId' -> Right txId'
+
+decodeAddressAny :: ByteString -> Either Text AddressAny
+decodeAddressAny address = case deserialiseFromRawBytes AsAddressAny address of
+  Nothing       -> Left "Invalid address"
+  Just address' -> Right address'
+
+-- GetIntersectionPoints
+
+getIntersectionPoints :: GetIntersectionPoints Session
+getIntersectionPoints = GetIntersectionPoints $ statement () $ rmap decodeResults
+  [vectorStatement|
+    SELECT slotNo :: bigint, id :: bytea
+    FROM   chain.block
+    WHERE  rollbackToBlock IS NULL
+    ORDER BY slotNo DESC LIMIT 2160
+  |]
+  where
+    decodeResults :: Vector (Int64, ByteString) -> [ChainPoint]
+    decodeResults = fmap decodeResult . V.toList
+
+    decodeResult :: (Int64, ByteString) -> ChainPoint
+    decodeResult (slotNo, hash) = ChainPoint (SlotNo $ fromIntegral slotNo) (HeaderHash $ toShort hash)
+
+-- CommitRollback
+
+commitRollback :: CommitRollback Session
+commitRollback = CommitRollback \case
+  ChainPointAtGenesis -> sql $ BS.intercalate "\n"
+    [ "BEGIN;"
+    , "TRUNCATE TABLE chain.tx;"
+    , "TRUNCATE TABLE chain.txOut;"
+    , "TRUNCATE TABLE chain.txIn;"
+    , "DELETE FROM chain.asset;"
```

Comment:
Why `DELETE` instead of `TRUNCATE`?

Reply:
Right, the assets table is a table that has those.

Reply:
You can use `TRUNCATE TABLE <table> CASCADE` for this actually.

I'm also pushing a fix that re-adds the genesis block after doing all this truncation... still faster to truncate and re-add than to exclude it using a `DELETE`